cpp/velox/compute/VeloxBackend.cc

```diff
@@ -342,19 +280,130 @@ void VeloxBackend::initCache(const std::unordered_map<std::string, std::string>&
               << ", IO threads: " << ioThreads;
   }
+
+  std::unordered_map<std::string, std::string> configurationValues;
+
+#ifdef ENABLE_S3
+  // yuan.zhou@intel.com
+  std::string awsAccessKey = conf.at("spark.hadoop.fs.s3a.access.key");
+  std::string awsSecretKey = conf.at("spark.hadoop.fs.s3a.secret.key");
+  std::string awsEndpoint = conf.at("spark.hadoop.fs.s3a.endpoint");
+  std::string sslEnabled = conf.at("spark.hadoop.fs.s3a.connection.ssl.enabled");
+  std::string pathStyleAccess = conf.at("spark.hadoop.fs.s3a.path.style.access");
+  std::string useInstanceCredentials = conf.at("spark.hadoop.fs.s3a.use.instance.credentials");
+  std::string iamRole = conf.at("spark.hadoop.fs.s3a.iam.role");
+  std::string iamRoleSessionName = conf.at("spark.hadoop.fs.s3a.iam.role.session.name");
+
+  const char* envAwsAccessKey = std::getenv("AWS_ACCESS_KEY_ID");
+  if (envAwsAccessKey != nullptr) {
+    awsAccessKey = std::string(envAwsAccessKey);
+  }
+  const char* envAwsSecretKey = std::getenv("AWS_SECRET_ACCESS_KEY");
+  if (envAwsSecretKey != nullptr) {
+    awsSecretKey = std::string(envAwsSecretKey);
+  }
+  const char* envAwsEndpoint = std::getenv("AWS_ENDPOINT");
+  if (envAwsEndpoint != nullptr) {
+    awsEndpoint = std::string(envAwsEndpoint);
+  }
+
+  std::unordered_map<std::string, std::string> s3Config({});
+  if (useInstanceCredentials == "true") {
+    s3Config.insert({
+        {"hive.s3.use-instance-credentials", useInstanceCredentials},
+    });
+  } else if (!iamRole.empty()) {
+    s3Config.insert({
+        {"hive.s3.iam-role", iamRole},
+    });
+    if (!iamRoleSessionName.empty()) {
+      s3Config.insert({
+          {"hive.s3.iam-role-session-name", iamRoleSessionName},
+      });
+    }
+  } else {
+    s3Config.insert({
+        {"hive.s3.aws-access-key", awsAccessKey},
+        {"hive.s3.aws-secret-key", awsSecretKey},
+    });
+  }
+  // Only need to set s3 endpoint when not use instance credentials.
+  if (useInstanceCredentials != "true") {
+    s3Config.insert({
+        {"hive.s3.endpoint", awsEndpoint},
+    });
+  }
+  s3Config.insert({
+      {"hive.s3.ssl.enabled", sslEnabled},
+      {"hive.s3.path-style-access", pathStyleAccess},
+  });
+
+  configurationValues.merge(s3Config);
+#endif
+
+  auto properties = std::make_shared<const velox::core::MemConfig>(configurationValues);
+
+  if (ioThreads > 0) {
+    ioExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(ioThreads);
+    FLAGS_split_preload_per_driver = splitPreloadPerDriver;
+    // Use global memory pool for hive connector if SplitPreloadPerDriver was enabled. Otherwise, the allocated memory
+    // blocks might cause unexpected behavior (e.g. crash) since the allocations were proceed in background IO threads.
+    velox::connector::registerConnector(std::make_shared<MemoryPoolReplacedHiveConnector>(
+        gluten::defaultLeafVeloxMemoryPool().get(), kHiveConnectorId, properties, ioExecutor_.get()));
+  } else {
```

Comment:
I prefer use global memory pool only when split preload enabled. IO executor does not always has problem.